include/exadg/convection_diffusion/driver.cpp

```diff
@@ -188,28 +187,152 @@ Driver<dim, Number>::ale_update() const
   time_int_bdf->ale_update();
 }
 
+template<int dim, typename Number>
+void
+Driver<dim, Number>::mark_cells_coarsening_and_refinement(dealii::Triangulation<dim> & tria,
+                                                          VectorType const & solution) const
+{
+  mark_cells_kelly_error_estimator(tria,
+                                   pde_operator->get_dof_handler(),
+                                   pde_operator->get_constraints(),
+                                   *pde_operator->get_mapping(),
+                                   solution,
+                                   application->get_parameters().degree +
+                                     1 /* n_face_quadrature_points */,
+                                   application->get_parameters().amr_data);
+}
+
+template<int dim, typename Number>
+void
+Driver<dim, Number>::setup_after_coarsening_and_refinement()
+{
+  pde_operator->initialize_dof_handler_and_constraints();
+  pde_operator->setup();
+
+  postprocessor->setup(*pde_operator);
+
+  VectorType const * velocity_ptr = nullptr;
+  VectorType         velocity;
+
+  if(application->get_parameters().problem_type == ProblemType::Unsteady)
+  {
+    if(application->get_parameters().temporal_discretization == TemporalDiscretization::BDF)
+    {
+      std::shared_ptr<TimeIntBDF<dim, Number>> time_integrator_bdf =
+        std::dynamic_pointer_cast<TimeIntBDF<dim, Number>>(time_integrator);
+
+      if(application->get_parameters().get_type_velocity_field() == TypeVelocityField::DoFVector)
+      {
+        Assert(false, dealii::ExcMessage("ALE and AMR combination not implemented."));
+        pde_operator->initialize_dof_vector_velocity(velocity);
+        pde_operator->interpolate_velocity(velocity, time_integrator->get_time());
+        velocity_ptr = &velocity;
+      }
+
+      auto & grid = *application->get_grid_non_const();
+      grid.coarse_triangulations =
+        dealii::MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          *application->get_grid_non_const()->triangulation);
```

Comment:
I do not like this design. The update of the coarse triangulations in case something has changed is the responsibility of the `Grid` class. So I believe we need to think more fundamentally about the design. As a rule, we should not allow to return non-const pointers to `Grid`.

Moreover, `MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence()` seems to be just one variant of our create-grid-functionality in ExaDG. Hence, this function call should in my opinion not appear on the driver level, but inside some other class/free function responsible for these grid-related functionalities.

Another design aspect is that this code should in my opinion not appear in `setup_after_coarsening_and_refinement()` but at the place where the fine triangulation is manipulated during AMR.

Reply:
this is still open! separate PR moving `grid` needed.

Reply:
Thank you for pointing this out. We are currently missing out on the mappings that are potentially present even if we do not employ ALE.
For now, I will first in a separate PR move the `grid` (and static mapping) to the `driver`, where the related ALE functionality already resides, and then come back here, apply changes to fit that approach and then introduce a member function in Grid, that updates the coarse triangulations and the respective mappings to be called from the driver like so:

`tria.execute_coarsening_and_refinement();`

`// TODO: update the mapping OR ale_mapping, both owned by the Driver`

`grid.update_coarse_triangulations_and_mappings(mapping);`

What we still miss then is the `TODO` mentioned here, which is easy to achieve for a mapping of non-`MappingQCache` type, otherwise I would get the displacement vector described by `MappingQCache` (new member function get_mapping_dof_vector or maybe even exists) and then interpolate this vector as well.

how does that sound?